Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h

```diff
@@ -0,0 +1,429 @@
+    template <class Graph>
+    struct Noborder_predicate;
+
+    struct Cluster_properties;
+    struct Decimation_properties;
+    
+    // typedefs
+    typedef typename GeomTraits::Point_3 Point_3;
+
+    typedef typename boost::graph_traits<TriangleMesh>::vertex_descriptor vertex_descriptor;
+    typedef typename boost::graph_traits<TriangleMesh>::edge_descriptor edge_descriptor;
+    typedef typename boost::graph_traits<TriangleMesh>::face_descriptor face_descriptor;
+
+    typedef CGAL::Dual<TriangleMesh> Dual_graph;
+    typedef boost::filtered_graph<Dual_graph, Noborder_predicate<TriangleMesh>> Filtered_dual_graph;
+    
+    typedef boost::property<cluster_props_t, Cluster_properties> VertexProperty;
+    typedef boost::property<decimation_props_t, Decimation_properties> EdgeProperty;
+
+    typedef boost::adjacency_list<boost::setS, boost::listS, boost::undirectedS, VertexProperty, EdgeProperty> Graph;
+    
+    typedef typename boost::graph_traits<Graph>::vertex_descriptor graph_vertex_descriptor;
+    typedef typename boost::graph_traits<Graph>::edge_descriptor graph_edge_descriptor;
+
+    typedef typename boost::property_map<Graph, cluster_props_t>::type Graph_cluster_map;
+    typedef typename boost::property_map<Graph, decimation_props_t>::type Graph_decimation_map;
+
+    // constants
+    const double CONCAVITY_FACTOR = 0.1;
+
+public:
+    Approx_decomposition(const TriangleMesh& mesh, const GeomTraits& traits)
+    : m_mesh(mesh)
+    , m_traits(traits)
+    , m_concavity_calc(mesh, traits)
+    {}
+
+    /**
+     * Computes approximate convex decomposition of a triangle mesh and fills up face property map with cluster-ids.
+     * @param face_ids which associates each face of a triangle mesh to a cluster-id [0, 'number_of_clusters'-1]
+     * @param concavity_threshold concavity value each cluster must satisfy
+     * @param min_number_of_clusters minimal number of cluster that can be produced, except the case when number of faces is less
+     */
+    template <class FacePropertyMap>
+    std::size_t decompose(FacePropertyMap face_ids, double concavity_threshold, std::size_t min_number_of_clusters)
+    {
+        // create filtered dual graph without border edges (null source or target vertex)
+        Dual_graph dual(m_mesh);
+        Filtered_dual_graph filtered_dual(dual, Noborder_predicate<TriangleMesh>(m_mesh));
+
+        // fill up adjacency list and its properties
+        setup_graph(filtered_dual);
+
+        // compute initial decimation costs and concavity values for all edges
+        BOOST_FOREACH(graph_edge_descriptor edge, edges(m_graph))
+        {
+            update_edge(edge, concavity_threshold, CONCAVITY_FACTOR);
+        }
+
+        // main loop that stops when all edges with concavities lower that `concavity_threshold` (valid edges)
+        // are decimated or the constraint of minimum number of cluster is met
+        while (num_vertices(m_graph) > min_number_of_clusters)
```

Comment:
We are looking for the minimum number of clusters that satisfy the `concavity_threshold`. On each iteration of the loop, num_vertices is decreased by one.

Reply:
At the beginning of the algorithm each face defines a cluster but you want to merge clusters so that the number is below a given bound which is the maximum number of threshold desired.

Reply:
Ok, I understand you.

Reply:
But no, we cannot gurantee the threshold of maximum number of clusters is met. When there is no valid edges to decimate anymore, it stops with more clusters than that threshold. I think too leave min_number of clusters and add precondition that the number of faces must be larger or equal to the threshold.